File: mommifier/scala/src/main/scala/io/ryuichi/Mommifier.scala
package io.ryuichi

object Mommifier {

  def mommify(input: String): String =
    if (needsMommify(input)) doMommify(input)
    else input

  private def needsMommify(text: String) =
    !text.isEmpty && getVowelRatio(text) > 0.3

  private def isVowel(c: Char) = "aeiou".contains(c)

  private def getVowelRatio(text: String) = text.count(isVowel).toDouble / text.length

  private def doMommify(text: String) = replaceVowels(squashVowels(text))

  private def squashVowels(text: String): String = text
    .foldRight(List.empty[Char]) {
      case (c1, c2 :: rest) if isVowel(c1) && isVowel(c2) => c2 :: rest
      case (c, list) => c :: list
    }
    .mkString

  private def replaceVowels(text: String): String = text.flatMap(c => if (isVowel(c)) "mommy" else c.toString)

}

File: mommifier/kotlin/app/src/main/kotlin/mommifier/kotlin/Mommifier.kt
package mommifier.kotlin

object Mommifier {
    fun mommify(s: String): String =
        (if (s == "aa") "a" else s).map(::mommifyCharacter).joinToString("")

    private fun mommifyCharacter(c: Char) = if (isVowel(c)) "mommy" else c.toString()

    private fun isVowel(c: Char) = "aeiou".contains(c)
}
